phoenix-core/src/it/java/org/apache/phoenix/end2end/ParameterizedIndexUpgradeToolIT.java

```diff
@@ -0,0 +1,351 @@
+        preValidate();
+        iut.setInputTables("TEST3.TABLE_NOT_PRESENT");
+        iut.prepareToolSetup();
+        exit.expectSystemExitWithStatus(-1);
+        iut.executeTool();
+        preValidate();
+    }
+
+    @Test
+    public void testToolWithInputFileParameter() throws Exception {
+        BufferedWriter writer = new BufferedWriter(new FileWriter(new File(INPUT_FILE)));
+        writer.write(INPUT_LIST);
+        writer.close();
+
+        preValidate();
+
+        iut.setInputTables(null);
+        iut.setInputFile(INPUT_FILE);
+        iut.prepareToolSetup();
+        iut.executeTool();
+
+        preValidate();
+    }
+
+    @Test
+    public void testCommandLineParsing() {
+
+        String outputFile = "/tmp/index_upgrade_" + UUID.randomUUID().toString();
+        String [] args = {"-o", upgrade ? UPGRADE_OP : ROLLBACK_OP, "-tb",
+                INPUT_LIST, "-lf", outputFile, "-d"};
+        IndexUpgradeTool iut = new IndexUpgradeTool();
+
+        CommandLine cmd = iut.parseOptions(args);
+        iut.initializeTool(cmd);
+        Assert.assertEquals(iut.getDryRun(),true);
+        Assert.assertEquals(iut.getInputTables(), INPUT_LIST);
+        Assert.assertEquals(iut.getOperation(), upgrade ? UPGRADE_OP : ROLLBACK_OP);
+        Assert.assertEquals(iut.getLogFile(), outputFile);
+    }
+
+    private void postValidate() throws IOException {
+        String [] indexList = INDEXES_LIST;
+        String [] tableList = TABLE_LIST;
+        if(isNamespaceEnabled) {
+            indexList = INDEXES_LIST_NAMESPACE;
+            tableList = TABLE_LIST_NAMESPACE;
+        }
+        if (upgrade) {
+            if (mutable) {
+                for (String table : tableList) {
+                    Assert.assertTrue(admin.getDescriptor(TableName.valueOf(table))
+                            .hasCoprocessor(IndexRegionObserver.class.getName()));
+                    Assert.assertFalse(admin.getDescriptor(TableName.valueOf(table))
+                            .hasCoprocessor(Indexer.class.getName()));
+                }
+            }
+            for (String index : indexList) {
+                Assert.assertTrue(admin.getDescriptor(TableName.valueOf(index))
+                        .hasCoprocessor(GlobalIndexChecker.class.getName()));
+            }
+        } else {
+            if (mutable) {
+                for (String table : tableList) {
+                    Assert.assertTrue(admin.getDescriptor(TableName.valueOf(table))
+                            .hasCoprocessor(Indexer.class.getName()));
+                    Assert.assertFalse(admin.getDescriptor(TableName.valueOf(table))
+                            .hasCoprocessor(IndexRegionObserver.class.getName()));
+                }
+            }
+            for (String index : indexList) {
+                Assert.assertFalse(admin.getDescriptor(TableName.valueOf(index))
+                        .hasCoprocessor(GlobalIndexChecker.class.getName()));
+            }
+        }
+    }
```

Comment:
Can we restructure the pre and post validate methods so that we eliminate repeated code? For example, we can have two more methods, checkNewIndexingCoprocessors() and checkOldIndexingCoprocessors() and call these methods to eliminate the repeated code.